packages/block-editor/src/hooks/use-bindings-attributes.js

```diff
@@ -56,181 +55,111 @@ export function canBindAttribute( blockName, attributeName ) {
- * @return {null}                              Data-handling component. Render nothing.
- */
-function BlockBindingBridge( { blockProps, bindings, onPropValueChange } ) {
-	const blockBindingsSources = unlock(
-		useSelect( blocksStore )
-	).getAllBlockBindingsSources();
+		const { setAttributes } = props;
 
-	return (
-		<>
-			{ Object.entries( bindings ).map(
-				( [ attrName, boundAttribute ] ) => {
-					// Bail early if the block doesn't have a valid source handler.
-					const source =
-						blockBindingsSources[ boundAttribute.source ];
-					if ( ! source?.useSource ) {
-						return null;
+		const _setAttributes = useCallback(
+			( nextAttributes ) => {
+				registry.batch( () => {
+					if ( ! bindings ) {
+						return setAttributes( nextAttributes );
 					}
 
-					return (
-						<BindingConnector
-							key={ attrName }
-							attrName={ attrName }
-							source={ source }
-							blockProps={ blockProps }
-							args={ boundAttribute.args }
-							onPropValueChange={ onPropValueChange }
-						/>
-					);
-				}
-			) }
-		</>
-	);
-}
-
-const withBlockBindingSupport = createHigherOrderComponent(
-	( BlockEdit ) => ( props ) => {
-		/*
-		 * Collect and update the bound attributes
-		 * in a separate state.
-		 */
-		const [ boundAttributes, setBoundAttributes ] = useState( {} );
-		const updateBoundAttributes = useCallback(
-			( newAttributes ) =>
-				setBoundAttributes( ( prev ) => ( {
-					...prev,
-					...newAttributes,
-				} ) ),
-			[]
-		);
+					const keptAttributes = { ...nextAttributes };
+
+					for ( const [
+						attributeName,
+						boundAttribute,
+					] of Object.entries( bindings ) ) {
+						const source = sources[ boundAttribute.source ];
+						if (
+							! source?.setValue ||
+							! canBindAttribute( name, attributeName )
+						) {
+							continue;
+						}
+
+						source.setValue( {
+							registry,
+							context,
+							clientId,
+							attributeName,
+							args: boundAttribute.args,
+							value: nextAttributes[ attributeName ],
+						} );
+						delete keptAttributes[ attributeName ];
+					}
 
-		/*
-		 * Create binding object filtering
-		 * only the attributes that can be bound.
-		 */
-		const bindings = Object.fromEntries(
-			Object.entries( props.attributes.metadata?.bindings || {} ).filter(
-				( [ attrName ] ) => canBindAttribute( props.name, attrName )
-			)
+					if ( Object.keys( keptAttributes ).length ) {
+						setAttributes( keptAttributes );
+					}
+				} );
+			},
+			[
+				registry,
+				bindings,
+				name,
+				clientId,
+				context,
+				setAttributes,
+				sources,
+			]
 		);
 
 		return (
 			<>
-				{ Object.keys( bindings ).length > 0 && (
-					<BlockBindingBridge
-						blockProps={ props }
-						bindings={ bindings }
-						onPropValueChange={ updateBoundAttributes }
-					/>
-				) }
-
 				<BlockEdit
 					{ ...props }
 					attributes={ { ...props.attributes, ...boundAttributes } }
+					setAttributes={ _setAttributes }
```

Comment:
Not sure if you have already been aware of this, but one reason we initially implemented Pattern Overrides differently is because patching `setAttributes` might not work in some edge cases.

There are different ways to update a block's attributes. Calling `setAttributes` from the `<BlockEdit>`'s props is just one of those ways. You can also call `updateBlockAttributes` directly from the `block-editor` store, `updateBlock` or `replaceBlock` to change the whole block, or even a `replaceInnerBlocks` and `cloneBlock` combo for whatever reasons.

We can make sure all core blocks favor `setAttributes` from the `<BlockEdit>`'s props whenever possible, but as we allow third-party blocks support, then it's probably going to be a problem.

A current example of this in the core block library would be the `core/table-of-contents` block. As you update the heading block elsewhere, then the ToC block will be updated using `updateBlockAttributes`.

The way we implemented in Pattern Overrides currently is the other way around. Instead of patching anything or "capturing" anything from the block updates, we "listen" to the changes that happened in the block and reflect the changes to the pattern instance (`content` attribute.) Everything that happens in the block is kept intact, so we don't have to patch anything. Hence, `syncDerivedUpdates` is used to "sync" the block's "updates" to the pattern instance. Actions wrapped in that action callback shouldn't create an undo level as it's merely syncing derived states between different blocks.

(Just for the info for https://github.com/WordPress/gutenberg/pull/60721)

Reply:
> There are different ways to update a block's attributes. Calling setAttributes from the <BlockEdit>'s props is just one of those ways. You can also call updateBlockAttributes directly from the block-editor store, updateBlock or replaceBlock to change the whole block, or even a replaceInnerBlocks and cloneBlock combo for whatever reasons.

That's a good point but it's not an issue that is specific to pattern overrides. All block bindings have the same issue/impact here.

Reply:
I'm aware of it and this is actually something that not using hooks would allow us to fix. But I remember @SantosGuillamot saying it was not desirable to fix, I can't recall correctly. Please correct me if I'm wrong 🙂

Reply:
I agree it is a good point and something we should aim to fix.

> But I remember @SantosGuillamot saying it was not desirable to fix, I can't recall correctly.

Mmm, I don't remember exactly. Maybe we were discussing the possibility of not using the hook and moving it to the core directly, and I mentioned that it shouldn't be a blocker for WordPress 6.6. But I totally agree that should probably be the path to go.